internal/widget/base.go

```diff
@@ -10,34 +12,72 @@ type base struct {
 	hidden bool
 	pos    fyne.Position
 	size   fyne.Size
+
+	propertyLock sync.RWMutex
 }
 
 // Move satisfies the fyne.Widget interface.
 func (b *base) Move(pos fyne.Position) {
-	b.pos = pos
+	b.setFieldsAndRefresh(nil, func() {
+		b.pos = pos
+	})
 }
 
 // Position satisfies the fyne.Widget interface.
 func (b *base) Position() fyne.Position {
-	return b.pos
+	var pos fyne.Position
+	b.readFields(func() {
+		pos = b.pos
+	})
+	return pos
+}
+
+// readFields provides a guaranteed thread safe way to access widget fields.
+func (b *base) readFields(f func()) {
+	b.propertyLock.RLock()
+	defer b.propertyLock.RUnlock()
+
+	f()
+}
+
+// SetFieldsAndRefresh helps to make changes to a widget that should be followed by a refresh.
+// This method is a guaranteed thread-safe way of directly manipulating widget fields.
+func (b *base) setFieldsAndRefresh(w fyne.Widget, f func()) {
```

Comment:
nit: the name ordering doesn't match parameter ordering.
```suggestion
func (b *base) setFieldsAndRefresh(f func(), w fyne.Widget) {
```